python/ray/experimental/gpu_buffer/torch_test.py

```diff
@@ -0,0 +1,100 @@
+import argparse
+import time
+
+import torch
+import torchvision
+import torchvision.transforms as transforms
+import torch.optim as optim
+import torch.nn as nn
+
+from models import SimpleCNN
+
+
+def download_dataset():
+    transform = transforms.Compose(
+    [transforms.ToTensor(),
+     transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+    trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
+    testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)
+    return trainset, testset
+
+def train_epoch(net, criterion, optimizer, trainloader, pinned):
+    net.train()
+    for i,data in enumerate(trainloader):
+        inputs, labels = data[0].to(device, non_blocking=pinned), data[1].to(device, non_blocking=pinned)
+        optimizer.zero_grad()
+        
+        out = net(inputs)
+        loss = criterion(out, labels)
+        loss.backward()
+        optimizer.step()
+
+def test(net, testloader, pinned):
+    net.eval()
+    total = 0
+    correct = 0
+    with torch.no_grad():
+        for data in testloader:
+            images, labels = data[0].to(device, non_blocking=pinned), data[1].to(device, non_blocking=pinned)
+            outputs = net(images)
+            _, predicted = torch.max(outputs.data, 1)
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+    return correct / total
```

Comment:
I would need to check this; line 42 might be done on CPU, which causes the GPU to block until the CPU computation is done. This would negate any benefits of nonblocking transfers.

Reply:
In general we want the CPU to run as far ahead as possible of the GPU, so that the GPU remains saturated.

Reply:
Using `tensor.get_device()`, I verified that `outputs`, `predicted`, and `labels` are on the GPU. Is that enough to conclude line 42 is done on GPU?

Reply:
[`.item()`](https://pytorch.org/docs/stable/generated/torch.Tensor.item.html) will convert the tensor value to a Python value, which blocks the CPU until the computation is complete